### PR TITLE
Issue 584/journey timeline framework

### LINKED ISSUE #584: Timeline framework components
## Description
All timeline events have some common elements, that could probably be implemented using some framework that concrete event types can then live within.

## Relevant Job Stories
See #572 

## Prerequisites
None

## Requirements
- [x] A timeline contains a list of events, separated by lines
- [x] Timeline events should contain an (optional) "actor" – the person who triggered the event
- [x] Timeline events should contain a (rich) text label
- [x] Timeline events should contain a timestamp, rendered as a relative date
- [x] Timeline events can contain complex content

In a future iterations, a timeline event will also have a star button for "bookmarking" and an ellipsis menu for further actions.

## Possible implementations
Use `ZetkinRelativeTime` for timestamp.

The API will return timeline events in a format similar to the one below (details pending), which would be rendered something like "**Clara Zetkin** assigned **Clara Zetkin** yesterday".

```js
{
  name: 'journeyinstance/addassignee',
  actor: {
    id: 123,
    first_name: 'Clara',
    last_name: 'Zetkin',
  },
  timestamp: '2022-02-22T19:17:00.000Z',
  target: {
    id: 1,
    title: "My case",
  },
  details: {
    // Data format varies by type
    assignee: {
      id: 123,
      first_name: 'Clara',
      last_name: 'Zetkin',
    }
  }
}
```

### Events with separators
The event ellipsis menu and star button are not part of this task.

image

### Minimal event with only common fields
image

### Event with complex content
image

## Open questions
None

### LINKED ISSUE #585: Timeline event: Journey assignee
## Description
A user should be able to see information in the timeline about journey instance assignees changing.

## Relevant Job Stories
* When a case is handed over to me, I want to be able to see who else has worked on the case, so that I know who to contact if I need any historical information

## Prerequisites
The API is still pending, but there should still be enough information to proceed.

## Requirements
- [x] Internationalized message when an assignee is removed
- [x] Internationalized message when an assignee is added
- [x] Relative timestamp
- [x] Hover card on people

## Possible implementations
API event object should look like this (TBD):

```js
{
  id: "123869812983619863-0",
  timestamp: "1857-07-05T12:34:56.000Z",
  name: "journeyinstance/addassignee", //or "removeassignee"
  target: {
    id: 1,
    title: "My Case",
  },
  actor: {
    id: 2,
    first_name: "Clara",
    last_name: "Zetkin",
  },
  details: {
    assignee: {
      id: 3,
      first_name: "Angela",
      last_name: "Davis",
    },
  },
}
```

## Design specifications
image

### PR DESCRIPTION
**Pull request by lowlandjuju**
## Description
Closes #584
Closes #585 

## Screenshots
issue-584

## Changes
- Adds ZetkinUpdate type 
- Adds Timeline and TimelineUpdate child component
- TimelineUpdate works for "assigned to" type
- Implements unit test for timeline component
- Adds storybook and several plugins as dev dependencies
- Adds a story for the timeline component
- Small optional UI modifier for ZetkinPerson avatar

## Notes to reviewer
- It looks like a lot of added lines, but most of those are in the yarn.lock file
- Leaving ZetkinUpdate type as hardcoded for "assigned by" update type for now. This can be extended to a more generic type when the second update type is added

### CODE REVIEW

**Review by richardolsson — COMMENTED** [review]
Here's some early feedback on this since we're working mostly async nowadays.

I really like that you introduced Storybook. I've never used it before, but it's been suggested before and I have been wanting to introduce it at the right time and this seems like a very good time.

I also really like the way the component looks. I wish I was able to style things with as much ease as you do.

My only concern has to do with the internationalization. I've elaborated on that in comments below. I've also updated the "needs-info" issue to reflect the direction that the data format has taken with my proof of concept.

Finally, a little tip. The PR template calls for issue references to be put at the bottom, and GitHub requires them to be individual lines, each with their own "Closes", "Resolves" or "Fixes" keyword, for automation to work. Note how issue #585 isn't being picked up by GitHub right now.

**Review by richardolsson — APPROVED** [review]
I think this looks great!

I don't mind merging it as is if that helps you progress. I do have a couple of questions below though that I think might require action, but again, it's not blocking as I foresee this code changing a bit over the next couple of PRs anyway.

File: src/components/PersonHoverCard/index.tsx [hunk 1]
```diff
@@ -10,10 +18,10 @@ import ZetkinPerson from 'components/ZetkinPerson';
 import { ZetkinPerson as ZetkinPersonType } from 'types/zetkin';
 import { personResource, personTagsResource } from 'api/people';
 
-const PersonHoverCard: React.FunctionComponent<{ personId: number }> = ({
-  children,
-  personId,
-}) => {
+const PersonHoverCard: React.FunctionComponent<{
+  BoxProps?: BoxProps;
```
[the remark this answers is not in the document]
Comment (lowlandjuju) [reply src/components/PersonHoverCard/index.tsx:22]: It's to distinguish between props supplied to virtual dom elements and actual dom elements.

File: src/components/Timeline/index.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  const [expanded, setExpanded] = React.useState<boolean>(!!showAll);
+
+  return (
+    <Fade appear in timeout={1000}>
+      <Grid container direction="column" spacing={3}>
+        {renderUpdateList()}
+        <Grid item>
+          <Button onClick={() => setExpanded(!expanded)} variant="outlined">
+            <FormattedMessage id="misc.timeline.expand" />
+          </Button>
```
[the remark this answers is not in the document]
Comment (lowlandjuju) [reply src/components/Timeline/index.tsx:28]: I saw it in the figma

Screenshot 2022-04-25 13 49 35

[the remark this answers is not in the document]
Comment (richardolsson) [reply src/components/Timeline/index.tsx:28]: I created #637 so we don't forget.

**Pull request merged**